Update user registration

NMertsch opened the pull request:
### Behavioral Changes
- Ticket name matching: Perform unicode normalization, try multiple name permutations (e.g. family name first)
- Pretix data fetching: Perform a full fetch on start, then periodically (attempt to) load updated orders every 5 minutes
- Registration: Make registration response message persistent for users in the #registration channel
- Registration log: `@tag` users in the #registration-log channel, list assigned roles by name

### Configuration Changes
- Update server, channel and role IDs in `config.toml`
- Simplify configuration: Remove `ticket_to_roles_(prod|staging).json`, set ticket-role-associations by role name

### Misc
- `pyproject.toml`: Use same line length for isort and black
- `.gitignore`: Add `.venv` and `registered_log.txt`
- `README.md`: Add setup instructions for Ubuntu (mainly because of pyenv, pipenv)

### Code structure
- Moved all registration-related code to a new `registration` subdirectory
- Reworked Pretix connector and registration code to increase maintainability and testability, and reduce effort for future handovers
- Separated registration tracking from Pretix API interactions
- Added tests for all parts but the Discord interactions

### Dependencies
- Added [unidecode](https://pypi.org/project/Unidecode/) for robust unicode normalization
- Added [pydantic](https://pypi.org/project/pydantic/) for parsing Pretix API responses

Kislovskiy commented:
This is awesome, thank you @NMertsch! Could you remind me how we are testing this?

NMertsch commented:
@Kislovskiy There were only very basic [tests for the pretix API interactions](https://github.com/EuroPython/discord/blob/6486913b7e84f7334e7e61438e6647db62b12a62/tests/test_pretix_connector.py#L25-L91), which I left untouched. I added [tests for name handling](https://github.com/EuroPython/discord/blob/6486913b7e84f7334e7e61438e6647db62b12a62/tests/test_pretix_connector.py#L94-L122).

You are right, there is much room for improvement. Working on it 👍

Kislovskiy commented:
@NMertsch  I was just curious if we have any test servers :)
If I remember correctly, the bot could start anywhere with the correct tokens provided.

NMertsch commented:
@Kislovskiy 
Yes, we have a test Discord server and test Pretix instance ("staging"). Please ping me on Discord if you want to have access to the instance and tokens.

I now added unit tests for all components besides the Discord Cog (and moved some responsibilities out of the `PretixConnector` class to make it easier to test in isolation).

I manually tested the four possible outcomes of submitting the registration form:
- Success
- Ticket already used
- Ticket not found
- Unexpected error

Here is a screenshot from the channel `#registration-log` of two users trying to register with the same ticket:

![image](https://github.com/EuroPython/discord/assets/49114330/66844387-82be-497e-9ad0-5d2ba46ec688)

The first user's nickname was correctly changed to the name on the ticket ("TODOG GODOT") and the user received the roles associated with the ticket type ("Onsite Speaker").

Kislovskiy commented on tests/registration/test_pretix_connector.py:
```diff
@@ -0,0 +1,272 @@
+    with pytest.raises(aiohttp.ClientResponseError) as e:
+        await pretix_connector.fetch_pretix_data()
+
+    assert e.value.code == 500
```
What about instead of `500`, use `HTTPStatus.INTERNAL_SERVER_ERROR` (like we do in `pretix_connector.py`)? That way, I won't need to Google what `500` stands for.

NMertsch replied:
Thank you, fixed in 091869e

Kislovskiy commented on EuroPythonBot/registration/cog.py:
```diff
@@ -0,0 +1,192 @@
+            '3️⃣ Click "Submit". We\'ll verify your ticket and assign you your roles based on '
+            "your ticket type.\n\n"
+            f"Experiencing trouble? Ask for help in the <#{config.REG_HELP_CHANNEL_ID}> channel "
+            "or from a volunteer in yellow t-shirt at the conference.\n\n"
```
```suggestion
            "or from a volunteer in a yellow t-shirt at the conference.\n\n"
```

NMertsch replied:
Thank you, fixed in 4e91759

Kislovskiy commented on EuroPythonBot/registration/pretix_connector.py:
```diff
@@ -0,0 +1,126 @@
+    def get_ticket(self, *, order: str, name: str) -> Ticket | None:
+        """Get the ticket for a given order ID and name, or None if none was found."""
+        # try different name orders (e.g. family name first vs last)
+        name_parts = name.split(maxsplit=5)
```
Could you also add why we have `maxsplit=5` here?

NMertsch replied:
Thank you, fixed in dc82da4

Kislovskiy commented on EuroPythonBot/registration/pretix_connector.py:
```diff
@@ -0,0 +1,126 @@
+
+            key = generate_ticket_key(order=order, name=possible_name)
+
+            if key in self.tickets_by_key:
+                return self.tickets_by_key[key]
+
+        return None
```
We could shrink it further like:
```suggestion
            return self.tickets_by_key.get(key, None)
```

Tested like this:

```Python
>>> a = {"A": 1, "B": 2}
>>> "A" in a
True
>>> a.get("A", None)
1
>>> a.get("C", None)
>>> res = a.get("C", None)
>>> res
>>> type(res)
<class 'NoneType'>
```

NMertsch replied:
We want to try all combinations until we find a match, and only return `None` if no combination matches. With your suggestion, we would return `None` if the first combination failed, without trying the other combinations.

Kislovskiy commented on EuroPythonBot/registration/pretix_connector.py:
```diff
@@ -0,0 +1,126 @@
+                _logger.debug("Fetching %s", url)
+
+                if next_url != url:
+                    params = None  # only send params on initial request
```
Oh, maybe let's put the comment before `if`. My tired brain was trying to understand what was going on here, and I didn't see the `# only send params on initial request` comment

NMertsch replied:
Good call, fixed in 35a6eb30

Kislovskiy commented on EuroPythonBot/registration/pretix_connector.py:
```diff
@@ -0,0 +1,126 @@
+                if next_url != url:
+                    params = None  # only send params on initial request
+
+                async with session.get(
+                    next_url, headers=self._http_headers, params=params
+                ) as response:
```
Maybe we could consider increasing the default length to something bigger than 100; there are multiple places where this line goes beyond one, which looks bad. What do you think?

NMertsch replied:
Thank you, I didn't see what black did to my code. Fixed in a42795ca by moving `headers` from request to session.

Kislovskiy reviewed (approved):
Very nice @NMertsch! Thanks a lot for adding many tests; they made the review effortless.
Please feel free to agree or disagree with the comments and merge the PR.
